micro-rdk-nmea/src/messages/pgns.rs

```diff
@@ -106,3 +115,96 @@ pub struct GnssSatsInView {
     #[length_field = "sats_in_view"]
     satellites: Vec<Satellite>,
 }
+
+macro_rules! define_pgns {
+    ( $(($pgndef:ident, $enum:ident, $pgn:expr)),* ) => {
+        #[derive(Clone, Debug)]
+        pub enum MessageData {
+            $($enum($pgndef)),*,
+            Unsupported(UnparsedMessageData)
+        }
+
+        impl MessageData {
+            pub fn pgn(&self) -> u32 {
+                match self {
+                    $(Self::$enum(_) => $pgn),*,
+                    Self::Unsupported(unparsed) => unparsed.pgn()
+                }
+            }
+
+            pub fn from_bytes(pgn: u32, bytes: Vec<u8>) -> Result<Self, crate::parse_helpers::errors::NmeaParseError> {
+                Ok(match pgn {
+                    $($pgn => {
+                        let cursor = DataCursor::new(bytes);
+                        Self::$enum($pgndef::from_cursor(cursor)?)
+                    }),*,
+                    x => Self::Unsupported(UnparsedMessageData::from_bytes(bytes, x)?)
+                })
+            }
+
+            pub fn to_readings(self) -> Result<GenericReadingsResult, crate::parse_helpers::errors::NmeaParseError> {
+                match self {
+                    $(Self::$enum(msg) => msg.to_readings()),*,
+                    Self::Unsupported(msg) => msg.to_readings()
+                }
+            }
+        }
+    };
+}
+
+define_pgns!(
+    (WaterDepth, Pgn128267, 128267),
```

Comment:
So, if I understand correctly, the generated enum will look like:
```
enum MessageData {
    Pgn128267(WaterDepth),
    Pgn130316(TemperatureExtendedRange),
    ...
}
```

So, do you actually need the second item in the tuple? 

```
define_pgns!(
    (WaterDepth, 127267),
    (TemperatureExtendedRange, 130316),
    ...
```
would seem to have all of the information required to generate the enum, and you wouldn't need to write the PGN value twice, once as part of a type name and the other as an integer.

Reply:
it's hard to interpolate tokens, I can try to see if the paste crate helps out here

Reply:
Sure, that sounds promising. But see my other comment just below about the naming of these enum members anyway. I'm not convinced that names with the PGN code in them are what is desired